### PR TITLE
clientv3: Do no stop keep alive loop by server side errors

### PR DESCRIPTION
**Pull request by yudai**
For #7880

### CODE REVIEW

File: clientv3/lease.go [hunk 1]
```diff
@@ -392,34 +392,46 @@ func (l *lessor) recvKeepAliveLoop() (gerr error) {
...
+			for {
+				resp, err := stream.Recv()
+
+				if err != nil {
+					if canceledByCaller(l.stopCtx, err) {
+						return err
+					}
+
+					err = toErr(l.stopCtx, err)
+					if err == rpctypes.ErrNoLeader {
```
Comment (yudai) [inline clientv3/lease.go]: Fixed

File: clientv3/client.go [hunk 2]
```diff
@@ -503,3 +503,16 @@ func toErr(ctx context.Context, err error) error {
...
 }
+
+func canceledByCaller(ctx context.Context, err error) bool {
+	if ctx.Err() == nil {
+		return false
+	}
+	if err == nil {
+		return false
+	}
+	if err != context.Canceled && err != context.DeadlineExceeded {
```
Comment (yudai) [inline clientv3/client.go]: Fixed. I just wanted to return `true` at the end of the function (now it's `false`, `false`, `true`, `false`).

Comment (heyitsanthony) [inline clientv3/client.go]: huh? the if isn't needed, the predicate can be returned directly..

Comment (yudai) [inline clientv3/client.go]: ah! sorry, I misunderstood your first comment. Yes, your way is much simpler and easy to understand :)

**Comment by heyitsanthony** [comment]
lgtm

File: clientv3/client.go [hunk 3]
```diff
@@ -503,3 +503,14 @@ func toErr(ctx context.Context, err error) error {
 	}
 	return err
 }
+
+func canceledByCaller(ctx context.Context, err error) bool {
+	if ctx.Err() == nil {
```
Comment (xiang90) [inline clientv3/client.go]:
``` go
if ctx.Err() == nil || err == nil {
    return false
}
```

Comment (xiang90) [inline clientv3/client.go]: probably also rename ctx to stopCtx.

Comment (yudai) [inline clientv3/client.go]: Thanks. Done.

**Comment by xiang90** [comment]
@yudai We probably want a test to cover this. It can be a different PR though. LGTM.

**Comment by yudai** [comment]
@xiang90 Adding a test could take a little bit time. Because I think we need a mock server that returns unexpected (not in `isHaltErr()`) errors.

**Comment by xiang90** [comment]
CI is fixed on master. Merging.

**Pull request merged**